File: app/src/main/java/com/github/premnirmal/ticker/portfolio/RearrangeAdapter.java
package com.github.premnirmal.ticker.portfolio;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.TextView;

import com.github.premnirmal.ticker.model.IStocksProvider;
import com.github.premnirmal.ticker.network.Stock;
import com.github.premnirmal.tickerwidget.R;
import com.nhaarman.listviewanimations.util.Swappable;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by premnirmal on 3/24/15.
 */
public class RearrangeAdapter extends BaseAdapter implements Swappable {

    private final List<Stock> stockList;
    private final IStocksProvider stocksProvider;

    RearrangeAdapter(IStocksProvider stocksProvider) {
        this.stocksProvider = stocksProvider;
        stockList = new ArrayList<>(this.stocksProvider.getStocks());
    }

    @Override
    public int getCount() {
        return stockList.size();
    }

    @Override
    public Stock getItem(int position) {
        return stockList.get(position);
    }

    @Override
    public long getItemId(int position) {
        return getItem(position).symbol.hashCode();
    }

    @Override
    public boolean hasStableIds() {
        return true;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        final Context context = parent.getContext();
        if(convertView == null) {
            convertView = LayoutInflater.from(context).inflate(R.layout.rearrange_view,null);
        }
        final Stock stock = getItem(position);
        ((TextView) convertView.findViewById(R.id.tickerName)).setText(stock.symbol + "\n(" + stock.Name + ")");
        return convertView;
    }

    @Override
    public void swapItems(int from, int to) {
        stockList.add(to, stockList.remove(from));
        final List<String> newTickerList = new ArrayList<>();
        for (Stock stock : stockList) {
            newTickerList.add(stock.symbol);
        }
        notifyDataSetChanged();
        stocksProvider.rearrange(newTickerList);
    }
}

File: app/src/main/kotlin/com/github/premnirmal/ticker/portfolio/drag_drop/RearrangeAdapter.kt
package com.github.premnirmal.ticker.portfolio.drag_drop

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.github.premnirmal.ticker.model.IStocksProvider
import com.github.premnirmal.ticker.network.data.Quote
import com.github.premnirmal.tickerwidget.R
import kotlinx.android.synthetic.main.rearrange_view.view.tickerName
import java.util.ArrayList

/**
 * Created by premnirmal on 2/29/16.
 */
internal class RearrangeAdapter internal constructor(private val stocksProvider: IStocksProvider,
    private val dragStartListener: OnStartDragListener)
  : RecyclerView.Adapter<ItemViewHolder>(), ItemTouchHelperAdapter {

  private val quoteList: MutableList<Quote>

  init {
    quoteList = ArrayList(this.stocksProvider.getStocks())
  }

  override fun getItemCount(): Int {
    return quoteList.size
  }

  fun getItem(position: Int): Quote {
    return quoteList[position]
  }

  override fun getItemId(position: Int): Long {
    return getItem(position).symbol.hashCode().toLong()
  }

  override fun onCreateViewHolder(parent: ViewGroup, itemType: Int): ItemViewHolder {
    val context = parent.context
    val view: View = LayoutInflater.from(context).inflate(R.layout.rearrange_view, null)
    return ItemViewHolder(view)
  }

  override fun onBindViewHolder(viewHolder: ItemViewHolder, position: Int) {
    val stock = getItem(position)
    viewHolder.itemView.tickerName.text = "${stock.symbol}\n(${stock.name})"
    viewHolder.itemView.setOnLongClickListener {
      dragStartListener.onStartDrag(viewHolder)
      true
    }
  }

  override fun onItemMove(fromPosition: Int, toPosition: Int): Boolean {
    quoteList.add(toPosition, quoteList.removeAt(fromPosition))
    val newTickerList = quoteList.mapTo(ArrayList<String>()) { it.symbol }
    stocksProvider.rearrange(newTickerList)
    notifyItemMoved(fromPosition, toPosition)
    return true
  }

  override fun onItemDismiss(position: Int) {

  }
}